lib/pharos/cluster_manager.rb

```diff
@@ -47,12 +47,17 @@ def validate
     def apply_phases
       apply_phase(Phases::ValidateHost, config.hosts, ssh: true, parallel: true)
       apply_phase(Phases::ConfigureHost, config.hosts, ssh: true, parallel: true)
+      apply_phase(Phases::ConfigureCfssl, config.etcd_hosts, ssh: true, parallel: true)
+      apply_phase(Phases::ConfigureEtcdCa, [config.etcd_host], ssh: true, parallel: false)
+      apply_phase(Phases::ConfigureEtcd, config.etcd_hosts, ssh: true, parallel: true)
 
-      apply_phase(Phases::ConfigureMaster, config.master_hosts, ssh: true, parallel: false)
+      apply_phase(Phases::ConfigureMaster, config.master_hosts, ssh: true, parallel: false, master: config.master_host)
+      apply_phase(Phases::ConfigureMaster, config.master_hosts, ssh: true, parallel: false, master: config.master_host)
       apply_phase(Phases::ConfigureKubelet, config.worker_hosts, ssh: true, parallel: true) # TODO: also run this phase in parallel for the master nodes, if not doing an upgrade?
-      apply_phase(Phases::ConfigureClient, config.master_hosts, ssh: true, parallel: true)
+      apply_phase(Phases::ConfigureClient, [config.master_host], ssh: true, parallel: false)
 
       # master is now configured and can be used
+      apply_phase(Phases::ConfigureClusterInfo, [config.master_host], ssh: false, parallel: false, master: config.master_host)
```

Comment:
This causes a race condition... `ConfigureMaster` sets master specific ip and it's fixed to localhost here. Not sure how we can actually fix it without updating `kubeadm`?

Reply:
I mean: if node joins to the cluster between these two phases it will be configured with single master ip.

Reply:
I don't understand, what race condition?

Reply:
But the `JoinNode` only happens later, well after the `Phases::ConfigureMaster` and `Phases::ConfigureClusterInfo`?

Reply:
True, I'm thinking use cases where nodes are joining via autoscaling group etc.

Reply:
Hmm.... probably only relevant for the `ConfigureMaster` upgrade case, where kubeadm might replace the `cluster-info` configmap?

Reply:
It looks like kubeadm generates the `cluster-info` configmap based on the `/etc/kubernetes/admin.conf`... so should we be using `api.advertiseAddress = http://localhost:6443`?

Reply:
I tried to set `api.advertiseAddress = 127.0.0.1` but it did not have any effect.

Reply:
Quick read of the kubeadm code says that it should end up using the `cfg.API.AdvertiseAddress` in both the `/etc/kubernetes/admin.conf` and thus the `cluster-info` configmap... need to test, I guess.

Reply:
I'll test asap, it would be much better way to solve this.

Reply:
Unfortunately it looks like kubeadm overrides `API.AdvertiseAddress: "127.0.0.1"` with the external address associated with the default route:

* https://github.com/kubernetes/kubernetes/blob/v1.9.6/cmd/kubeadm/app/util/config/masterconfig.go#L40-L42
* https://github.com/kubernetes/apimachinery/blob/895b82eadff27f5a6b19d80628a04645823c1e46/pkg/util/net/interface.go#L384

I don't understand why, though... if you're binding to `127.0.0.1`, then the only address that you can connect to is also going to be `127.0.0.1`... it doesn't make any sense to me to default to the external IP address in that case.

The only other caller of `utilnet.ChooseBindAddress` seems to be for `kubeapiserver --advertise-address=0.0.0.0 --insecure-bind-address=127.0.0.1`... in that case it would chose the external IP as the advertise address, but if you pass `--advertise-address=127.0.0.1` then it would respect that: https://github.com/kubernetes/kubernetes/blob/v1.9.6/pkg/kubeapiserver/options/serving.go#L57

Feels like a bug in kubeadm's use of `netutil.ChooseBindAddress` to me...